consul/resource_consul_service.go

```diff
@@ -78,80 +128,165 @@ func resourceConsulServiceCreate(d *schema.ResourceData, meta interface{}) error
 		for i, raw := range vs {
 			s[i] = raw.(string)
 		}
-		registration.Tags = s
+		registration.Service.Tags = s
+	}
+
+	if _, err := catalog.Register(registration, &wOpts); err != nil {
+		return fmt.Errorf("Failed to register service (dc: '%s'): %v", dc, err)
+	}
+
+	service, err := retrieveService(client, name, ident, node, dc)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(service.ServiceID)
+
+	return resourceConsulServiceRead(d, meta)
+}
+
+func resourceConsulServiceUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*consulapi.Client)
+	catalog := client.Catalog()
+
+	name := d.Get("name").(string)
+	node := d.Get("node").(string)
+
+	dc := ""
+	if _, ok := d.GetOk("datacenter"); ok {
+		dc = d.Get("datacenter").(string)
 	}
 
-	if err := agent.ServiceRegister(&registration); err != nil {
-		return fmt.Errorf("Failed to register service '%s' with Consul agent: %v", name, err)
+	// Setup the operations using the datacenter
+	wOpts := consulapi.WriteOptions{Datacenter: dc}
+
+	registration := &consulapi.CatalogRegistration{
+		Datacenter: dc,
+		Node:       node,
+		Service: &consulapi.AgentService{
+			Service: name,
+		},
 	}
 
-	// Update the resource
-	if serviceMap, err := agent.Services(); err != nil {
-		return fmt.Errorf("Failed to read services from Consul agent: %v", err)
-	} else if service, ok := serviceMap[identifier]; !ok {
-		return fmt.Errorf("Failed to read service '%s' from Consul agent: %v", identifier, err)
+	// If we have a service_id
+	if serviceID, ok := d.GetOk("service_id"); ok {
+		registration.Service.ID = serviceID.(string)
+	}
+
+	if address, ok := d.GetOk("address"); ok {
+		registration.Address = address.(string)
+		registration.Service.Address = address.(string)
 	} else {
-		d.SetId(service.ID)
-
-		d.Set("address", service.Address)
-		d.Set("service_id", service.ID)
-		d.Set("name", service.Service)
-		d.Set("port", service.Port)
-		tags := make([]string, 0, len(service.Tags))
-		for _, tag := range service.Tags {
-			tags = append(tags, tag)
+		// If we don't have an address, skip updating the node
+		registration.SkipNodeUpdate = true
+	}
+
+	if port, ok := d.GetOk("port"); ok {
+		registration.Service.Port = port.(int)
+	}
+
+	if v, ok := d.GetOk("tags"); ok {
+		vs := v.([]interface{})
+		s := make([]string, len(vs))
+		for i, raw := range vs {
+			s[i] = raw.(string)
 		}
-		d.Set("tags", tags)
+		registration.Service.Tags = s
 	}
 
-	return nil
+	if _, err := catalog.Register(registration, &wOpts); err != nil {
+		return fmt.Errorf("Failed to update service (dc: '%s'): %v", dc, err)
+	}
+
+	return resourceConsulServiceRead(d, meta)
 }
 
 func resourceConsulServiceRead(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*consulapi.Client)
-	agent := client.Agent()
 
+	dc := ""
+	if _, ok := d.GetOk("datacenter"); ok {
+		dc = d.Get("datacenter").(string)
+	}
```

Comment:
Got it. DC is more of a query parameter that could be provided at the provider level. The handling of DC in both ways is pretty challenging to wrap your head around currently, so honestly I want to revisit that down the road. I'd rather _not_ be putting it into the ID now.

I tried to avoid the ID changing so that you could use this in a backwards compatible way by changing resource names and resources in the state file (see upgrade guide).

Reply:
👍 Just wanted to call it out. I think you could also do some trickery with import to get around this without changing the ID. So all in all, if this is how you want it, I'm 👍 with that.